src/transformations/transform-constructor.ts

```diff
@@ -78,86 +81,110 @@ export function* removeLeftoverConstructorHead(sourceUnit: SourceUnit): Generato
+    const { resolver, getData } = tools;
 
-  for (const contractNode of findAll('ContractDefinition', sourceUnit)) {
-    if (contractNode.contractKind !== 'contract' || hasConstructorOverride(contractNode)) {
-      continue;
-    }
+    const useNamespaces = isNamespaced?.(sourceUnit.absolutePath) ?? false;
+
+    for (const contractNode of findAll('ContractDefinition', sourceUnit)) {
+      if (contractNode.contractKind !== 'contract' || hasConstructorOverride(contractNode)) {
+        continue;
+      }
 
-    const { name } = contractNode;
-    const {
-      constructorNode,
-      varInitNodes,
-      modifiers,
-      emptyUnchained: emptyConstructor,
-    } = getInitializerItems(contractNode, resolver);
-
-    const initializer = (
-      helper: TransformHelper,
-      argsList = '',
-      unchainedArgsList = '',
-      argNames: string[] = [],
-    ) => [
-      `function __${name}_init(${argsList}) internal onlyInitializing {`,
-      buildSuperCallsForChain(contractNode, tools, helper),
-      emptyConstructor ? [] : [`__${name}_init_unchained(${argNames.join(', ')});`],
-      `}`,
-      ``,
-      [
-        `function`,
-        `__${name}_init_unchained(${unchainedArgsList})`,
-        `internal onlyInitializing`,
-        ...modifiers.map(m => helper.read(m)),
-        `{`,
-      ].join(' '),
-      varInitNodes.map(v => {
-        const newExpr = parseNewExpression(v.value!);
-        if (!newExpr) {
-          return `${v.name} = ${helper.read(v.value!)};`;
-        } else {
-          const { typeName, newCall, initializeCall } = newExpr;
-          const createdContract = resolver.resolveContract(typeName.referencedDeclaration);
-          if (createdContract) {
-            getData(createdContract).isUsedInNewStatement = true;
+      const { name } = contractNode;
+      const {
+        constructorNode,
+        varInitNodes,
+        modifiers,
+        emptyUnchained: emptyConstructor,
+      } = getInitializerItems(contractNode, resolver);
+
+      const namespace = getNamespaceStructName(name);
+      const constructorUsesStorage =
+        constructorNode !== undefined && usesStorageVariables(constructorNode, resolver);
+
+      const initializer = (
+        helper: TransformHelper,
+        argsList = '',
+        unchainedArgsList = '',
+        argNames: string[] = [],
+      ) => [
+        `function __${name}_init(${argsList}) internal onlyInitializing {`,
+        buildSuperCallsForChain(contractNode, tools, helper),
+        emptyConstructor ? [] : [`__${name}_init_unchained(${argNames.join(', ')});`],
+        `}`,
+        ``,
+        [
+          `function`,
+          `__${name}_init_unchained(${unchainedArgsList})`,
+          `internal onlyInitializing`,
+          ...modifiers.map(m => helper.read(m)),
+          `{`,
+        ].join(' '),
+        useNamespaces && varInitNodes.length > 0 && !constructorUsesStorage
```

Comment:
It took me time to understand that this helper for the case `varInitNodes` must be initialized in the initialized is only needed if `!constructorUsesStorage` because otherwize, `$` would already be declared as part of the constructor rewrite.

Maybe that should be documented by a comment

Reply:
This will not work, because if `constructorUsesStorage` is true, then the declaration of `$` will be after the `varInitNodes`. See test C8 I just added.

Reply:
Fixed in https://github.com/OpenZeppelin/openzeppelin-transpiler/pull/127/commits/f6deaa0abef025d3706f2b5e6c4539dbe16b22b7